adapters/loyal/loyal.go

```diff
@@ -0,0 +1,158 @@
+package loyal
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+type reqBodyExt struct {
+	LoyalBidderExt reqBodyExtBidder `json:"bidder"`
+}
+
+type reqBodyExtBidder struct {
+	Type        string `json:"type"`
+	PlacementID string `json:"placementId,omitempty"`
+	EndpointID  string `json:"endpointId,omitempty"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var err error
+	var adapterRequests []*adapters.RequestData
+
+	reqCopy := *request
+	for _, imp := range request.Imp {
+		reqCopy.Imp = []openrtb2.Imp{imp}
+
+		var bidderExt adapters.ExtImpBidder
+		var loyalExt openrtb_ext.ImpExtLoyal
+
+		if err = json.Unmarshal(reqCopy.Imp[0].Ext, &bidderExt); err != nil {
+			return nil, []error{err}
+		}
+		if err = json.Unmarshal(bidderExt.Bidder, &loyalExt); err != nil {
+			return nil, []error{err}
+		}
+
+		impExt := reqBodyExt{LoyalBidderExt: reqBodyExtBidder{}}
+
+		if loyalExt.PlacementID != "" {
+			impExt.LoyalBidderExt.PlacementID = loyalExt.PlacementID
+			impExt.LoyalBidderExt.Type = "publisher"
+		} else if loyalExt.EndpointID != "" {
+			impExt.LoyalBidderExt.EndpointID = loyalExt.EndpointID
+			impExt.LoyalBidderExt.Type = "network"
+		} else {
+			continue
+		}
```

Comment:
as per bidder-info yaml, either `loyalExt.PlacementID` or `loyalExt.EndpointID` are required and prebid server core framework will ensure eithe PlacementID or EndpointID value is present. Therefore `else` on line 62 should not be present

Reply:
done